Rewrite

rvolosatovs opened the pull request:
- The utilities now operate on actual, final field paths.
~Clients are expected to clean and convert the paths to final(it's quite trivial given that we know all the paths now), however we may want to delegate this to the `SetFields` methods themselves. Not sure whose responsibility should this be, hence did not implement that.~
```proto
message Test {
  message TestNested {
    bytes a = 1;
    int32 b = 2;
  }
  TestNested a = 1;
  fixed32 b = 2;
}

// Field paths are:
//  - a
//  - a.a
//  - a.b
//  - b
```

- ~`SetFields` method performs deep copy *always* and never copies the reference~
- ~For maps, customtypes, some well-known types and all message types we don't know about copying is done using reflection for now. Non-reflective copying for customtypes and maps is on the way(did not have enough time to finish that).~



NOTE: We need `gogoproto` protocol buffers in order to generate the extensions, however `go mod` is not too friendly about it for now, hence we need `dep` for (clean) builds.
The package is still perfectly `go get`'able and `go test`'able without `dep`.

NOTE2: I think it makes most sense to just check out the code locally and review it that way, treat it as a new project. There's not much similarity between the current `master` and this implementation.

@KrishnaIyer please do a first round of review and then let @htdvisser take a final look

rvolosatovs commented:
@KrishnaIyer will take over until I'm back on Monday

johanstokking commented:
@rvolosatovs there are 0 issues in this repository. PRs must refer to or close issues (https://thethingsindustries.slack.com/archives/C06UQSMPB/p1539094821000100). I'll come up with an updated template (in `lorawan-stack`) to propose how it should be implemented, to avoid lengthy and unexpected rewrites.

KrishnaIyer commented on main.go:
```diff
@@ -1,196 +1,861 @@
+	imports["fmt"] = "fmt"
+	fmt.Fprintf(buf, `
+		default:
+			panic(fmt.Errorf("Invaild fieldpath: '%%s'", path))
```
Not sure if `panic` is a good idea during setting the fields. Though the fieldmasks will be validated in the API boundaries and it's unlikely that an invalid field will get through that,  I think an `interal error` with cause should be sufficient.

rvolosatovs replied:
@htdvisser ?

KrishnaIyer commented on main.go:
```diff
@@ -1,196 +1,861 @@
+	for _, p := range paths {
+		fmt.Fprintf(buf, `
+		case "%s":
+			var nilPath bool`,
```
the variable `nilPath` is set only in L520 for OneOfs. I assume that the go complier is smart enough to remove this during optimization but it's redundant for non-oneof fields.

rvolosatovs replied:
`nilPath` is not set on `L520` - it's only evaluated there, it is set at `L258` and `L277`, however.
As you can see, it not redundant for non-oneof fields.
It is only redundant for field paths, which do not contain pointers in any of the links.

rvolosatovs replied:
Because if we're setting `A.B.C` we don't care if `A.D` is `nil`
`nilPath` only concerns the intermidiary steps in the path(so for `A.B.C.D` `nilPath == true` implies that one of `{,A,A.B,A.B.C}` is `nil`)

KrishnaIyer replied:
I also agree with @htdvisser. Since the scope of the variable is only within the the case, in the above snippet it's redundant  and the `if nilpath {}` is dead code that I assume that the go compiler will remove. You could define this variable above the switch and reset upon entering a branch.

rvolosatovs commented on main.go:
```diff
@@ -1,196 +1,861 @@
+	}
+}
+
+type plugin struct{}
```
I disagree. This change would mean that we essentially just rename `main.go` to `plugin.go`, as we end up with `main.go`, containing 5 lines(which is too short) and `plugin.go` containing the rest.

rvolosatovs commented on Makefile:
```diff
@@ -1,11 +1,44 @@
+	CGO_ENABLED=0 go build -ldflags "-w -s" -o dist/protoc-gen-fieldmask-$(shell go env GOOS)-$(shell go env GOARCH)$(shell go env GOEXE) .
 
-.PHONY: build
+WORKDIR:=$(shell mkdir -p $(PWD)/.work && mktemp -d -p $(PWD)/.work)
```
Please verify that the recent changes work on OSX

KrishnaIyer replied:
Checked this and it works.

johanstokking commented:
@rvolosatovs just for transparency and because we can't look in your head (fortunately!), can you walk us through the process of what's now in `master` to this rewrite? Specifically what is missing from there, and why the missing things couldn't be implemented there so that a rewrite is necessary, and that this is a lengthy process and produces lengthy code?

This is part of V3, and the goal there has always been to keep it simple. Things we elaborate on are the API, the architecture, etc. This is typical utility that should just work, according to the requirements in the bottom of https://github.com/TheThingsIndustries/lorawan-stack/issues/1212. I understand the devil is in the details, but again a bit of thought process helps.

rvolosatovs commented:
@htdvisser: The grow in the amount of lines is simply because now the fieldmask utilities are generated for all protos, instead of only the ones, for which a message containing a fieldmask field exists. We want this, because we want to be able to deep-copy types without reflection.

The solution in `master` simply does not work for our use case.
1. `master` solution copies value references - that is incorrect, as the solution should perform a deep copy

2. non-top level fieldpaths are unsupported and it's not straight-forward to add support for them __correctly__ (in fact, I tried to do that but eventually I ended up here)

First, for the rest of this message, assume source `src` and destination `dst`.

Consider path `A.B.C.D` is to be set. Distinguish cases:
   1. One of the links in path `src.A.B.C.D` is `nil` (one of `src.{,A,A.B,A.B.C}`) and all links in path `dst.A.B.C.D` are non-`nil` -> set `dst.A.B.C.D` to zero value
   2. One of the links in path `src.A.B.C.D` is `nil` (one of `src.{,A,A.B,A.B.C}`) and one of links in path `dst.A.B.C.D` is `nil` (one of `dst.{,A,A.B,A.B.C}`) -> don't do anything
   3. All of the links in path `src.A.B.C.D` are non-`nil` and all links in path `dst.A.B.C.D` are non-`nil` -> deep copy `src.A.B.C.D` to `dst.A.B.C.D`
   4. All of the links in path `src.A.B.C.D` are non-`nil` and one of links in path `dst.A.B.C.D` is `nil` -> initialize `dst.A.B.C.D` and deep copy `src.A.B.C.D` to `dst.A.B.C.D`


To illustrate the shortcomings let's assume that the support to `master` is added via using a `switch` on the prefix up to the dot. Consider message definition from PR description and fieldmask `a.b`, where `src.a == nil`. `master` solution will set `dst.a` to `nil`, which is invalid, as that will also change the value of `a.a`.

So issues with `master` assuming nested fieldpath support:
- Fields, for which a fieldmask is not specified may be deleted (really bad)
- A bunch of garbage empty fields may be initialized on delete (less severe)

Note, that solution in this branch using `nilPath` addresses exactly both of these shortcomings.

Regarding complexity IMO the code generated by the plugin in this branch is much easier to follow, as it operates directly on the final paths of each message, so per each path you can immediately see what happens without having to follow a long callstack.

htdvisser commented:
> The grow in the amount of lines is simply because now the fieldmask utilities are generated for all protos, instead of only the ones, for which a message containing a fieldmask field exists.

This is not true. The `master` branch generates utilities for all messages, regardless of whether they contain a field mask or not.

htdvisser commented:
> The grow in the amount of lines is simply because now the fieldmask utilities are generated for all protos, instead of only the ones, for which a message containing a fieldmask field exists.

This is not true. The `master` branch generates utilities for all messages, regardless of whether they contain a field mask or not.

> `master` solution copies value references - that is incorrect, as the solution should perform a deep copy

Then this is incorrect too:

```go
case "recent_downlinks":
	var nilPath bool
	if nilPath || src.RecentDownlinks == nil {
		dst.RecentDownlinks = nil
		continue
	}
	dst.RecentDownlinks = make([]*DownlinkMessage, len(src.RecentDownlinks))
	for i, v := range src.RecentDownlinks {
		dst.RecentDownlinks[i] = v
	}
```

> non-top level fieldpaths are unsupported and it's not straight-forward to add support for them correctly

That can easily be fixed by splitting the path on `.` for the `switch` in `master`. The rest of the func (so setting the field with `FieldMaskWithoutPrefix(mask, "field_name")` works just fine.

> Consider message definition from PR description and fieldmask a.b, where src.a == nil. master solution will set dst.a to nil, which is invalid, as that will also change the value of a.a.

This is simply an invalid situation. I don't think we should consider that.

>  I also came up with more robust/extendable solutions for the hacky workarounds, which currently exist in master

Those could have easily been added separately.

KrishnaIyer commented on main.go:
```diff
@@ -1,196 +1,864 @@
+				srcPath = fmt.Sprintf("%s.Get%s()", srcPath, goName)
+
+				fmt.Fprintf(buf, `
+			nilPath = nilPath || %s == nil`,
```
Could you please use brackets to make this and `L275` more readable?:
`nilPath = (nilPath || (%s == nil))`

rvolosatovs replied:
No. `nilPath = nilPath || %s == nil` is much more readable than `nilPath = (nilPath || (%s == nil))`, but if @johanstokking agrees, I'll change that.

rvolosatovs commented:
Implemented the cosmetic changes in generated code(at the expense of additional complexity) as requested by @htdvisser and @KrishnaIyer 
Current output for `lorawan-stack` can be seen at https://github.com/TheThingsIndustries/lorawan-stack/commit/93d9338be2a96932ea203a63848ae251db28d612

KrishnaIyer commented:
Nice! Apart from the comment on panics and the cosmetic one on brackets for readability, I'm good with this PR. Please get the opinions of @johanstokking and @htdvisser before merging.

johanstokking commented:
I'm gonna close this for now based on the previous discussion.

rvolosatovs commented:
I removed the deep copying and deleting of fields.
The implementation now simply panics with a NPE if the(intermediary) field in `dst` is `nil` and it's up to the caller to ensure it does not happen.
`FieldMaskPaths()` returns field paths of all levels of nesting.

I think we should not rely on `SetFields` when setting nested messages, setting fields directly is much clearer to follow/debug, simpler and is even more efficient.
For example see https://github.com/TheThingsIndustries/lorawan-stack/commit/80f24446e909a14d9dd11dc1d58e868a9b5247ef#diff-d564ed2d8b392d89a85c41644b5dda3eR36 it's short and concise and it's immediately obvious what happens. If the field had nesting of 20 levels deep, sure it would not look so good anymore, but in practice it does not happen and we rarely(if, at all) have messages with more than 3 levels of nesting.

The code in it's current state fulfills all the requirements of MVP with various improvements compared to `master`. Implementation with methods may look nicer, but this is generated code - looking nice is not a goal here.

htdvisser commented:
> I think we should not rely on SetFields when setting nested messages, setting fields directly is much clearer to follow/debug, simpler and is even more efficient.
For example see TheThingsIndustries/lorawan-stack@80f2444#diff-d564ed2d8b392d89a85c41644b5dda3eR36 it's short and concise and it's immediately obvious what happens. If the field had nesting of 20 levels deep, sure it would not look so good anymore, but in practice it does not happen and we rarely(if, at all) have messages with more than 3 levels of nesting.

I'm sorry, but I'm just not going to approve this. The generated Unmarshal funcs also don't unmarshal sub-messages, but simply delegate that to the Unmarshal func of the sub-message.

> Implementation with methods may look nicer, but this is generated code - looking nice is not a goal here.

We should keep our code as concise as possible, even if it's generated. I don't want to end up with a 100MB binary just because we stop caring about generated code.

Setting a field of a KeyEnvelope deeply nested in the EndDevice is not the concern of the EndDevice, it's the concern of the KeyEnvelope.

rvolosatovs commented:
I changed the implementation to use methods, see output for `lorawan-stack` here: https://github.com/TheThingsIndustries/lorawan-stack/commit/c0d2e24aa4251e9faddacc11efda36f7ce96fb5e
I also added some cosmetic changes @johanstokking asked for.

The biggest difference from `master` now is that this version operates on the full recursive set of paths.

htdvisser commented on main.go:
```diff
@@ -1,196 +1,625 @@
+		if sp := strings.Split(p, "."); len(sp) > 1 {
+			fd := md.GetMessageField(sp[0])
+			if fd.GetLabel() == descriptor.FieldDescriptorProto_LABEL_REPEATED {
+				panic(errors.New("Fieldmask for repeated field generated"))
```
our errors are lowercase

htdvisser commented on main.go:
```diff
@@ -1,196 +1,625 @@
+				descriptor.FieldDescriptorProto_TYPE_FIXED64,
+				descriptor.FieldDescriptorProto_TYPE_STRING,
+				descriptor.FieldDescriptorProto_TYPE_BYTES:
+				return nil, fmt.Errorf("Invalid type specifed for embedded field: %s", fd.GetType())
```
1. our errors are lowercase
2. probably good to mention the containing message here instead of having something like `Invalid type specifed for embedded field: pointer`.

htdvisser commented on main.go:
```diff
@@ -1,196 +1,625 @@
-	messages[md.FullName] = md
+	fmt.Fprintf(buf, `
+		default:
+			panic(fmt.Errorf("Invalid field path: '%%s'", path))
```
our errors are lowercase

rvolosatovs commented:
@htdvisser we can always change this later.
When implemented properly, adding a new field will only result in 1 added word to the switch cases containing the submessage, which is totally fine IMO.
If we want to change the semantics and instead operate on the prefixes - we can easily do that later. The implementation is very flexible. But the point is that we have an MVP, which works and is simple to maintain/extend/change and generates simple code.

rvolosatovs commented:
With "proper" implementation I mean https://github.com/TheThingsIndustries/protoc-gen-fieldmask/issues/7, but note that the same thing could also be done with prefix matching

rvolosatovs commented:
That is indeed the case with current implementation. See https://github.com/TheThingsIndustries/lorawan-stack/commit/71c8331f6b97bdb2928a3baacaca6d1b51837da4

htdvisser commented:
I was referring to 

```go
case "end_device.pending_session.keys.s_nwk_s_int_key.kek_label":
	dst.EndDevice.SetFields(&src.EndDevice, _pathsWithoutPrefix("end_device", paths)...)
```

rvolosatovs commented:
It does not - it delegates this to the next subfield in exactly the same way `master` does it, except that we match the full path here instead of using a prefix matching.

rvolosatovs commented:
PR in `lorawan-stack` using this functionality https://github.com/TheThingsIndustries/lorawan-stack/pull/1316

rvolosatovs commented:
We can change this in #7 and use prefix matching. It's a very simple and straight-forward change, but it is not a priority right now.
